standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -1006,16 +1007,23 @@ public boolean checkFailedCompactions(CompactionInfo ci) throws MetaException {
         rs = pStmt.executeQuery();
         int numFailed = 0;
         int numTotal = 0;
+        long lastEnqueueTime = -1;
         int failedThreshold = MetastoreConf.getIntVar(conf, ConfVars.COMPACTOR_INITIATOR_FAILED_THRESHOLD);
         while(rs.next() && ++numTotal <= failedThreshold) {
+          long enqueueTime = rs.getLong(2);
+          if (enqueueTime > lastEnqueueTime) {
+            lastEnqueueTime = enqueueTime;
+          }
           if(rs.getString(1).charAt(0) == FAILED_STATE) {
             numFailed++;
           }
           else {
             numFailed--;
           }
         }
-        return numFailed == failedThreshold;
+        // If the last attempt was too long ago, ignore the failed treshold and try compaction again
+        long retryTime = MetastoreConf.getTimeVar(conf, ConfVars.COMPACTOR_INITIATOR_FAILED_RETRY_TIME, TimeUnit.MILLISECONDS);
```

Comment:
The config's value is retrieved every time each table/partition is checked. Might be better to get it in the main loop of Initiator instead.

Reply:
I don't really want to do that. The same is true for COMPACTOR_INITIATOR_FAILED_THRESHOLD a few lines up, but I think this makes this code more readable and more testable by its own. Besides this part is not that time critical, since it is a background process.

Reply:
I'm not 100% convinced about the time critical part, since effectively, because of mutexing, only one Initiator thread runs on the whole system.

Reply:
Still, this does not initiate anything in the conf, it is already there in memory. Just run some test on my machine the average time for a getTimeVar call is 0.48 microsecond